code/datums/immutable_list.dm

```diff
@@ -0,0 +1,137 @@
+///Real globals have no datum access overhead, so we do both to maintain GLOB having all the important bits.
+GLOBAL_REAL(_immutable_list_repo, /list) = list()
+GLOBAL_LIST_INIT(immutable_list_repo, global._immutable_list_repo)
+
+/datum/immutable_list
+	VAR_FINAL/list/data
+
+/datum/immutable_list/New(list/_data)
+	data = _data
+
+///Constructor for immutable lists, use the macro IMMUTABLE_LIST().
+/proc/__immutable_list(list/data)
+	var/list/refs = list()
+	for(var/argument in data)
+		refs += ref(argument) //Ignore datum tags, we don't care.
+
+	var/key = refs.Join()
```

Comment:
you need a separator for this join, or you're gonna potentially have collisions (in the far future case where people actually use tags)

Reply:
bet